data/api/client-server/pusher.yaml

```diff
@@ -209,14 +214,15 @@ paths:
               lang:
                 type: string
                 description: |-
-                  The preferred language for receiving notifications (e.g. 'en'
-                  or 'en-US').
+                  Required if `kind` is not `null`. The preferred language for
+                  receiving notifications (e.g. 'en' or 'en-US').
               data:
                 type: object
                 description: |-
-                  A dictionary of information for the pusher implementation
-                  itself. If `kind` is `http`, this should contain `url`
-                  which is the URL to use to send notifications to.
+                  Required if `kind` is not `null`. A dictionary of information
+                  for the pusher implementation itself. If `kind` is `http`,
+                  this should contain `url` which is the URL to use to send
+                  notifications to.
```

Comment:
I'm not sure this is necessarily required in principle, although in practice all (both) types of pusher need some information here.

Reply:
Synapse requires the field to be present: https://github.com/matrix-org/synapse/blob/develop/synapse/rest/client/pusher.py#L95-L106

Is there really something required for an `email` pusher? The spec doesn't say so.

Reply:
Does that mean that it should also not be required on the `GET /_matrix/client/v3/pushers` endpoint?

Reply:
Hmm - probably: I'd expect most to include it anyway to guard against lazy clients, but that doesn't mean the spec should force them to. What do others think?

Reply:
Mostly, I think that making `data` optional here is even more of a divergence from the established spec, and harder to justify without an MSC. We can make an exception if `kind` is `null`, because (a) it's clearly nonsense to require a `data` in that case and (b) that's what synapse has done forever, but going beyond that seems a stretch.

If we want to relax the requirements for `data` when `kind` is *not* `null`, let's make it a separate change.